### PR TITLE
Make storage tests work in parallel

### PR DESCRIPTION
**Pull request by rmohr**
**What this PR does / why we need it**:

In order to make our storage tests work in parallel, our NFS server and our local storage PVCs and PVs need to work properly in the parallel setup. Since the storage tests had a lot of custom waiting and checking related to cleanup, this PR also adds some matchers and improves parts which go a little bejond making storage run in parallel, to make it easier to grasp what the storage tests do.

**New PV and PVC management**

* Our local PVs for cirros and alpine now provide three copies on each node, so that we can run three tests in parallel without having to think about PV/PVC reuse.
* The local storage PVs and PVCs will be regenerated after each test
* The NFS-ganesha server can now be run in parallel for each test.
* All generated PVs and PVCs are now clearly bound to a specific test namespace
* Cleanup of PVCs, PVs and Datavolumes happens now automatically based on the clear binding to a test namespace

**New Matchers**

The storage tests do a lot of custom waiting for various things. To reduce the boilerplate and help the refactoring to parallel execution, a set of new matchers was added the `tests/framework` package. It allows matching like this:

Wait for single objects or sets of objects to disappear:
```go
Eventually(ThisVMI(vmi), 120).Should(BeGone())
Eventually(ThisPod(pod), 120).Should(Not(Exist()))
Eventually(ThisDV(datavolume), 120).Should(BeGone())
Eventually(AllVMIs("mynamespace"), 120).Should(BeGone())
```

Ensuring that objects don't disappear:
```go
Consistently(ThisVMI(vmi), 120).Should(Exist())
Consistently(ThisPod(pod), 120).Should(BePresent())
Consistently(ThisDV(datavolume), 120).Should(Exist())
```

Checking once if objects exist or disappear:
```go
Expect(ThisVMI(vmi), 120).To(BeGone())
Expect(ThisPod(pod), 120).To(Not(Exist()))
Expect(ThisPod(pod), 120).To(BePresent())
```

Checking if objects have reached certain phases:
```go
Eventually(ThisVMI(vmi), 120).Should(BeInPhase("Running"))
Eventually(ThisPod(pod), 120).Should(Not(HavePhase("Running")))
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

This PR does everything necessary to make storage tests work in parallel and adds some matchers to make it easier to grasp how the storage tests work, but it does **not yet** really clean up our test setup. In a followup we can introduce a clear test context which has all these loosely coupled functions properly structured and can provide a concise context to a test.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: tests/framework/matcher/phase.go
Comment (fgimenez) [inline tests/framework/matcher/phase.go:85]: Do we need to return an error here?

File: tests/framework/matcher/phase_test.go
Comment (fgimenez) [inline tests/framework/matcher/phase_test.go:28]: Maybe we could add an additional case for unexisting actual phase (ie. `runtime.DefaultUnstructuredConverter.ToUnstructured` or `unstructured.NestedString` returning an error), wdyt?

File: tests/utils.go
Comment (fgimenez) [inline tests/utils.go:1654]:
```suggestion
			if err != nil && !errors.IsNotFound(err) {
				PanicOnError(err)
			}
```

File: tests/utils.go
Comment (fgimenez) [inline tests/utils.go:991]:
```suggestion
			Name: Sprintf("%s-%s", name, NamespaceTestDefault),
```

**Review by fgimenez — COMMENTED** [review]
Just a bunch of nitpicks :)

File: tests/framework/storage/nfs.go
Comment (rmohr) [reply tests/framework/storage/nfs.go:17]: Indeed it is. Great catch.

File: tests/framework/matcher/existence.go
Comment (rmohr) [reply tests/framework/matcher/existence.go:32]: Let me rewrite this function a little. But in general it catches two things: If `nil` is passed in, the `actual==nil` check matches. If something like `var pod *v1.Pod = nil` is passed in, then the `reflect.ValueOf(actual).IsNil()` check matches..

File: tests/framework/cleanup/cleanup.go
Comment (rmohr) [reply tests/framework/cleanup/cleanup.go:1]: No, that's not needed.

File: tests/framework/matcher/phase.go
Comment (rmohr) [reply tests/framework/matcher/phase.go]: Dropped the error and added more tests.

File: tests/utils.go
Comment (rmohr) [reply tests/utils.go]: Done.

File: tests/utils.go
Comment (rmohr) [reply tests/utils.go]: Done.

File: tests/framework/matcher/phase_test.go
Comment (rmohr) [reply tests/framework/matcher/phase_test.go:33]: Done.

File: tests/framework/matcher/existence.go
Comment (rmohr) [reply tests/framework/matcher/existence.go]: Refactored the function and added more tests which trigger all conditions.

**Comment by rmohr** [comment]
@fgimenez I think I addresse your comments. Happy to get another review :)

File: tests/framework/matcher/phase.go
Comment (fgimenez) [inline tests/framework/matcher/phase.go:80]: <3

File: tests/storage_test.go
Comment (fgimenez) [inline tests/storage_test.go:465]: Is it Without or With feature gates disabled?

File: tests/storage_test.go
Comment (fgimenez) [inline tests/storage_test.go:582]:
```suggestion
						diskName = fmt.Sprintf("disk-%s.img", uuid.NewRandom())
```

**Review by fgimenez — COMMENTED** [review]
Thanks! A couple more suggestions

File: tests/framework/matcher/phase_test.go
Comment (davidvossel) [inline tests/framework/matcher/phase_test.go:10]: this is a unit test. right? The unit tests only execute if they are a part of this

```
bazel test \
    --config=${ARCHITECTURE} \
    --stamp \                                                                                                                  
    --test_output=errors -- //staging/src/kubevirt.io/client-go/... //pkg/... //cmd/...
```

this dir isn't listed. So does this unit test get executed as part of our functional tests?

File: images/disks-images-provider/entrypoint.sh
Comment (davidvossel) [inline images/disks-images-provider/entrypoint.sh:51]: is there any practical way to parameterize the number of workers so we don't get surprised in the future by this when someone changes the test worker count?

File: tests/framework/matcher/phase_test.go
Comment (rmohr) [reply tests/framework/matcher/phase_test.go:10]: It is listed there: https://github.com/kubevirt/kubevirt/pull/4616/files#diff-c5d77fa510a26d6656083dd099dd8512f416243de60780c49c60d1c9ca2c3e7a

But it probably gets executed too on the functests. Need to fix that.

File: images/disks-images-provider/entrypoint.sh
Comment (rmohr) [reply images/disks-images-provider/entrypoint.sh:51]: I can check it. We may be able to inject an environment var with the number of workers. Need to reexamine the exact setup flow.The mist tricky part is, that this is done on `cluster-sync`, whereas the number of workers can change on every `make functest`.

File: tests/framework/matcher/phase_test.go
Comment (rmohr) [reply tests/framework/matcher/phase_test.go:10]: No, all good there. They are executed in the unit test phase as expected, and not executed for e2e tests, since we don't compile them into the test binary.

**Comment by oshoval** [comment]
Please see if sriov fails due to this PR, as it say something about CDI
and we don't have CDI on sriov.

https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4616/pull-kubevirt-e2e-kind-1.17-sriov/1334166281955315712

https://prow.apps.ovirt.org/view/gcs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4616/pull-kubevirt-e2e-kind-1.17-sriov/1334091548127662082

File: tests/storage_test.go
Comment (rmohr) [reply tests/storage_test.go]: Fixed.

File: tests/storage_test.go
Comment (rmohr) [reply tests/storage_test.go]: Fixed.

**Comment by rmohr** [comment]
> Please see if sriov fails due to this PR, as it say something about CDI
> and we don't have CDI on sriov.

Fixed

**Comment by rmohr** [comment]
After my refactoring, tests which were restarting VMIs multiple times ran into

```
vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43
Unexpected Warning event received: testvmi-k2wwmf99nxtm,8dbe46b9-564e-4e2f-bb30-53a77f2d82d7: unknown error encountered sending command SyncVMI: rpc error: code = DeadlineExceeded desc = context deadline exceeded
Expected
    <string>: Warning
not to equal
    <string>: Warning
tests/utils.go:360
```

sometimes. It turns out that this error was coming from previous starts where due to timing such a stale grpc connect can happen and is fine. I adjusted the watcher in the tests to not react anymore to these.

**Review by davidvossel — COMMENTED** [review]
/approve

File: images/disks-images-provider/entrypoint.sh
Comment (rmohr) [reply images/disks-images-provider/entrypoint.sh:51]: Saving this task for later. We will not increase it short-term and it is pretty hard to change in the scope of this PR.

**Comment by rmohr** [comment]
/hold

I have no explanation why the operators appear to be more flaky now (they run long after serial tests, and not even other lanes with more serial tests are running there).

**Comment by rmohr** [comment]
/hold cancel

Finally good. @davidvossel @fgimenez would be happy about another review.

File: tests/operator_test.go
Comment (davidvossel) [inline tests/operator_test.go:266]: for us to get here, this pod that's set for deletion still has containers with the containerStatus.Ready=true and the pod itself is phase=running.

couldn't this could be an issue for leader election if this pod is the leader?

**Review by davidvossel — COMMENTED** [review]
the pr looks great. is there any more info about what's going on with the operator functional tests. one of your changes made it look like we're stuck waiting on a pod to terminate. do you know which pod is causing issues? I'm not necessarily opposed to the change, I'm just hoping to understand what you were fighting there a little better.

File: tests/operator_test.go
Comment (rmohr) [reply tests/operator_test.go:266]: I could not really find out what was causing this. The only explanation I have is test order change due to making some tests parallel executable. The interesting part is that I only run these tests after serial tests are done (on all lanes, since they start roughly at the same time). My refactoring changes nothing in the operator test part and there should not be more load than before.

**Review by davidvossel — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
